File: dsa-interview/java/Heap.java
package java;

import java.util.*;

/**
 * Heap / Priority Queue Implementation
 * 
 * Common Interview Problems:
 * - Kth Largest Element (Leetcode 215)
 * - Top K Frequent Elements (Leetcode 347)
 * - Merge K Sorted Lists (Leetcode 23)
 * - Find Median from Data Stream (Leetcode 295)
 * - Task Scheduler (Leetcode 621)
 */
public class Heap {
    
    private int[] heap;
    private int size;
    private boolean isMinHeap;
    
    public Heap(int capacity, boolean isMinHeap) {
        this.heap = new int[capacity];
        this.size = 0;
        this.isMinHeap = isMinHeap;
    }
    
    private int parent(int i) { return (i - 1) / 2; }
    private int leftChild(int i) { return 2 * i + 1; }
    private int rightChild(int i) { return 2 * i + 2; }
    
    private boolean compare(int a, int b) {
        return isMinHeap ? a < b : a > b;
    }
    
    public void insert(int val) {
        if (size == heap.length) {
            heap = Arrays.copyOf(heap, heap.length * 2);
        }
        heap[size] = val;
        siftUp(size);
        size++;
    }
    
    private void siftUp(int i) {
        while (i > 0 && compare(heap[i], heap[parent(i)])) {
            swap(i, parent(i));
            i = parent(i);
        }
    }
    
    public int peek() {
        if (size == 0) throw new NoSuchElementException();
        return heap[0];
    }
    
    public int poll() {
        if (size == 0) throw new NoSuchElementException();
        int result = heap[0];
        heap[0] = heap[size - 1];
        size--;
        siftDown(0);
        return result;
    }
    
    private void siftDown(int i) {
        int best = i;
        int left = leftChild(i);
        int right = rightChild(i);
        
        if (left < size && compare(heap[left], heap[best])) {
            best = left;
        }
        if (right < size && compare(heap[right], heap[best])) {
            best = right;
        }
        
        if (best != i) {
            swap(i, best);
            siftDown(best);
        }
    }
    
    private void swap(int i, int j) {
        int temp = heap[i];
        heap[i] = heap[j];
        heap[j] = temp;
    }
    
    public int size() { return size; }
    public boolean isEmpty() { return size == 0; }
    
    // ==================== DEMO ====================
    public static void main(String[] args) {
        // Min Heap
        Heap minHeap = new Heap(10, true);
        minHeap.insert(5);
        minHeap.insert(3);
        minHeap.insert(8);
        minHeap.insert(1);
        
        System.out.println("Min Heap extraction:");
        while (!minHeap.isEmpty()) {
            System.out.print(minHeap.poll() + " ");  // 1 3 5 8
        }
        System.out.println();
        
        // Find Median
        System.out.println("\n--- Median Finder ---");
        MedianFinder mf = new MedianFinder();
        mf.addNum(1);
        mf.addNum(2);
        System.out.println("Median after [1,2]: " + mf.findMedian());
        mf.addNum(3);
        System.out.println("Median after [1,2,3]: " + mf.findMedian());
        
        // Top K Frequent
        System.out.println("\n--- Top K Frequent ---");
        int[] nums = {1, 1, 1, 2, 2, 3};
        System.out.println("Top 2 frequent: " + Arrays.toString(topKFrequent(nums, 2)));
    }
    
    // Kth Largest Element
    public static int findKthLargest(int[] nums, int k) {
        PriorityQueue<Integer> minHeap = new PriorityQueue<>();
        for (int num : nums) {
            minHeap.offer(num);
            if (minHeap.size() > k) {
                minHeap.poll();
            }
        }
        return minHeap.peek();
    }
    
    // Top K Frequent Elements
    public static int[] topKFrequent(int[] nums, int k) {
        Map<Integer, Integer> freq = new HashMap<>();
        for (int num : nums) {
            freq.put(num, freq.getOrDefault(num, 0) + 1);
        }
        
        PriorityQueue<Integer> minHeap = new PriorityQueue<>(
            (a, b) -> freq.get(a) - freq.get(b)
        );
        
        for (int num : freq.keySet()) {
            minHeap.offer(num);
            if (minHeap.size() > k) {
                minHeap.poll();
            }
        }
        
        int[] result = new int[k];
        for (int i = k - 1; i >= 0; i--) {
            result[i] = minHeap.poll();
        }
        return result;
    }
}

/**
 * Find Median from Data Stream (Leetcode 295)
 */
class MedianFinder {
    private PriorityQueue<Integer> maxHeap;  // Lower half
    private PriorityQueue<Integer> minHeap;  // Upper half
    
    public MedianFinder() {
        maxHeap = new PriorityQueue<>(Collections.reverseOrder());
        minHeap = new PriorityQueue<>();
    }
    
    public void addNum(int num) {
        maxHeap.offer(num);
        minHeap.offer(maxHeap.poll());
        
        if (minHeap.size() > maxHeap.size()) {
            maxHeap.offer(minHeap.poll());
        }
    }
    
    public double findMedian() {
        if (maxHeap.size() > minHeap.size()) {
            return maxHeap.peek();
        }
        return (maxHeap.peek() + minHeap.peek()) / 2.0;
    }
}

File: dsa-interview/cpp/Heap.cpp
/**
 * Heap / Priority Queue Implementation in C++
 */

#include <iostream>
#include <vector>
#include <queue>
#include <unordered_map>
#include <algorithm>

template<typename T, typename Compare = std::less<T>>
class Heap {
private:
    std::vector<T> heap;
    Compare comp;
    
    int parent(int i) { return (i - 1) / 2; }
    int leftChild(int i) { return 2 * i + 1; }
    int rightChild(int i) { return 2 * i + 2; }
    
    void siftUp(int i) {
        while (i > 0 && comp(heap[i], heap[parent(i)])) {
            std::swap(heap[i], heap[parent(i)]);
            i = parent(i);
        }
    }
    
    void siftDown(int i) {
        int best = i;
        int left = leftChild(i);
        int right = rightChild(i);
        
        if (left < heap.size() && comp(heap[left], heap[best])) {
            best = left;
        }
        if (right < heap.size() && comp(heap[right], heap[best])) {
            best = right;
        }
        
        if (best != i) {
            std::swap(heap[i], heap[best]);
            siftDown(best);
        }
    }

public:
    void push(T val) {
        heap.push_back(val);
        siftUp(heap.size() - 1);
    }
    
    T top() const {
        return heap[0];
    }
    
    void pop() {
        heap[0] = heap.back();
        heap.pop_back();
        if (!heap.empty()) siftDown(0);
    }
    
    int size() const { return heap.size(); }
    bool empty() const { return heap.empty(); }
};

// Find Median from Data Stream
class MedianFinder {
private:
    std::priority_queue<int> maxHeap;  // Lower half
    std::priority_queue<int, std::vector<int>, std::greater<int>> minHeap;  // Upper half

public:
    void addNum(int num) {
        maxHeap.push(num);
        minHeap.push(maxHeap.top());
        maxHeap.pop();
        
        if (minHeap.size() > maxHeap.size()) {
            maxHeap.push(minHeap.top());
            minHeap.pop();
        }
    }
    
    double findMedian() {
        if (maxHeap.size() > minHeap.size()) {
            return maxHeap.top();
        }
        return (maxHeap.top() + minHeap.top()) / 2.0;
    }
};

// Top K Frequent Elements
std::vector<int> topKFrequent(std::vector<int>& nums, int k) {
    std::unordered_map<int, int> freq;
    for (int num : nums) freq[num]++;
    
    auto cmp = [&freq](int a, int b) { return freq[a] > freq[b]; };
    std::priority_queue<int, std::vector<int>, decltype(cmp)> minHeap(cmp);
    
    for (auto& [num, count] : freq) {
        minHeap.push(num);
        if (minHeap.size() > k) minHeap.pop();
    }
    
    std::vector<int> result;
    while (!minHeap.empty()) {
        result.push_back(minHeap.top());
        minHeap.pop();
    }
    return result;
}

// Kth Largest Element
int findKthLargest(std::vector<int>& nums, int k) {
    std::priority_queue<int, std::vector<int>, std::greater<int>> minHeap;
    for (int num : nums) {
        minHeap.push(num);
        if (minHeap.size() > k) minHeap.pop();
    }
    return minHeap.top();
}

int main() {
    // Custom Heap
    Heap<int> minHeap;
    minHeap.push(5);
    minHeap.push(3);
    minHeap.push(8);
    minHeap.push(1);
    
    std::cout << "Min Heap extraction: ";
    while (!minHeap.empty()) {
        std::cout << minHeap.top() << " ";
        minHeap.pop();
    }
    std::cout << std::endl;
    
    // Median Finder
    std::cout << "\n--- Median Finder ---\n";
    MedianFinder mf;
    mf.addNum(1);
    mf.addNum(2);
    std::cout << "Median after [1,2]: " << mf.findMedian() << std::endl;
    mf.addNum(3);
    std::cout << "Median after [1,2,3]: " << mf.findMedian() << std::endl;
    
    // Top K Frequent
    std::cout << "\n--- Top K Frequent ---\n";
    std::vector<int> nums = {1, 1, 1, 2, 2, 3};
    auto result = topKFrequent(nums, 2);
    std::cout << "Top 2 frequent: ";
    for (int x : result) std::cout << x << " ";
    std::cout << std::endl;
    
    return 0;
}
